File: data/raw/eggAnalysis.m
% eggAnalysis.m - The purpose of this code is to perform adequate
% processing of EGG signal, to calculate dominant frequencies (DF), and to
% perform statistical analysis.
%
% INPUT:    EGG signals recorded in 20 subjects (.txt file)
% OUTPUT:   dominant frequencies (.csv file)
%
%   Written by Nenad B. Popovic (nenad.pop92@gmail.com) and
%   Nadica Miljkovic (nadica.miljkovic@etf.rs)
% 
% If you use this code, please, cite relevant paper and dataset as:
%   [1] Popovic, N.B., Miljkovic, N. and Popovic M.B., 2019. Simple gastric
%   motility assessment method with a single-channel electrogastrogram.
%   Biomedical Engineering/Biomedizinsche Technik, 64(2), pp.177-185, doi:
%   10.1515/bmt-2017-0218
%   [2] Popovic, N.B., Miljkovic, N. and Popovic M.B., 2020. Three-channel
%   surface electrogastrogram (EGG) dataset recorded during fasting and
%   post-prandial states in 20 heatlhy individuals [Data set]. Zenodo, doi:
%   10.5281/zenodo.3730617.
%
% GNU Octave, version 5.2.0
% Copyright (C) 2020 John W. Eaton and others.
% This is free software; see the source code for copying conditions.
% -------------------------------------------------------------------------

close all; clear all; clc;

% load Octave packages
pkg load signal; pkg load io; pkg load nan; % comment if you use Matlab

%% dominant peak analysis

% specify parameters
fs = 2; % Hz, sampling frequency
N = 4096; % number of points for FFT analysis

% memory allocation for Dominant Frequency (DF) calculation
df = zeros(20, 6);

% bandpass filter for noise reduction
[b, a] = butter(3, [0.03 0.25] / (fs / 2), 'bandpass');

% DF calculation
for ind = 1 : 20
    % FASTING -------------------------------------------------------------
    
    % signal loading
    file_name_f = ['ID' num2str(ind) '_fasting.txt'];
    dat_f = load(file_name_f);
    
    % EGG filtering
    ch1_f = filtfilt(b, a, dat_f(1:2400, 1)); % channel 1
    ch2_f = filtfilt(b, a, dat_f(1:2400, 2)); % channel 2
    ch3_f = filtfilt(b, a, dat_f(1:2400, 3)); % channel 3
    
    % FFT
    fch1_f = abs(fft(ch1_f, N)).^2; fftch1_f = fch1_f(1:(N/2+1));
    fch2_f = abs(fft(ch2_f, N)).^2; fftch2_f = fch2_f(1:(N/2+1));
    fch3_f = abs(fft(ch3_f, N)).^2; fftch3_f = fch3_f(1:(N/2+1));
    
    % automatic DF calculation
    [~, b1_f] = max(fftch1_f); df(ind, 1) = b1_f / 2048;
    [~, b2_f] = max(fftch2_f); df(ind, 2) = b2_f / 2048;
    [~, b3_f] = max(fftch3_f); df(ind, 3) = b3_f / 2048;
    
    % POSTPRANDIAL --------------------------------------------------------
    
    % signal loading
    file_name_p = ['ID' num2str(ind) '_postprandial.txt'];
    dat_p = load(file_name_p);

    % EGG filtering
    ch1_p = filtfilt(b, a, dat_p(1:2400, 1)); % channel 1
    ch2_p = filtfilt(b, a, dat_p(1:2400, 2)); % channel 2
    ch3_p = filtfilt(b, a, dat_p(1:2400, 3)); % channel 3
    
    % FFT
    fch1_p = abs(fft(ch1_p, N)).^2; fftch1_p = fch1_p(1:(N/2+1));
    fch2_p = abs(fft(ch2_p, N)).^2; fftch2_p = fch2_p(1:(N/2+1));
    fch3_p = abs(fft(ch3_p, N)).^2; fftch3_p = fch3_p(1:(N/2+1));
    
    % automatic DF calculation
    [~, b1_p] = max(fftch1_p); df(ind, 4) = b1_p / 2048;
    [~, b2_p] = max(fftch2_p); df(ind, 5) = b2_p / 2048;
    [~, b3_p] = max(fftch3_p); df(ind, 6) = b3_p / 2048;  
    
end

% conversion from Hz to cpm (cycles per minute)
df = df.*60;

% all peaks were individually inspected by visualization and manual
% corrections were provided for false DF values in order to provide true
% dominant peaks
df(4, 4) = 3.1934; % ID4 postprandial channel 1
df(4, 5) = 3.1348; % ID4 postprandial channel 2
df(4, 6) = 3.1348; % ID4 postprandial channel 3
df(6, 3) = 2.4900; % ID6 fasting channel 3
df(15, 4) = 2.2560; % ID15 postprandial channel 1
df(17, 1) = 2.9592; % ID17 fasting channel 1
df(17, 2) = 3.0469; % ID17 fasting channel 2
df(17, 3) = 3.0762; % ID17 fasting channel 3

%% saving DF values in a file
filename = 'df.csv';
csvwrite(filename, df)

%% paired-sample t-tests

% for all subjects ID1-ID20
[h1, p1] = ttest( df(:, 1), df(:, 4) );
[h2, p2] = ttest( df(:, 2), df(:, 5) );
[h3, p3] = ttest( df(:, 3), df(:, 6) );

% for subjects with lower body mass index (BMI)
df_low(1:10, :) = [df(1:3, :); df(9, :); df(12, :); df(14, :); df(16:19, :)];
[h4, p4] = ttest( df_low(:, 1), df_low(:, 4) );
[h5, p5] = ttest( df_low(:, 2), df_low(:, 5) );
[h6, p6] = ttest( df_low(:, 3), df_low(:, 6) );

% for subjects ID11-ID20 (higher BMI)
df_high(1:10, :) = [df(4:8, :); df(10:11, :); df(13, :); df(15, :); df(20, :)];
[h7, p7] = ttest( df_high(:, 1), df_high(:, 4) );
[h8, p8] = ttest( df_high(:, 2), df_high(:, 5) );
[h9, p9] = ttest( df_high(:, 3), df_high(:, 6) );

% show p-values (p) and test decisions (h)
p = [p1, p2, p3, p4, p5, p6, p7, p8, p9]
h = [h1, h2, h3, h4, h5, h6, h7, h8, h9]

File: data/raw/egg_analysis.py
import numpy as np
import pandas as pd
from scipy.signal import butter, filtfilt
from scipy.fft import fft
from scipy.stats import ttest_rel
from sklearn.model_selection import train_test_split

# specify parameters
fs = 2  # Hz, sampling frequency
N = 4096  # number of points for FFT analysis

# memory allocation for Dominant Frequency (DF) calculation
df = np.zeros((40, 6))  # Updated to 40 entries

# bandpass filter for noise reduction
b, a = butter(3, [0.03, 0.25], btype='bandpass', fs=fs)

# DF calculation for existing 20 participants
for ind in range(1, 21):
    # FASTING -------------------------------------------------------------
    
    # signal loading
    file_name_f = f'ID{ind}_fasting.txt'
    dat_f = np.loadtxt(file_name_f)
    
    # EGG filtering
    ch1_f = filtfilt(b, a, dat_f[:2400, 0])  # channel 1
    ch2_f = filtfilt(b, a, dat_f[:2400, 1])  # channel 2
    ch3_f = filtfilt(b, a, dat_f[:2400, 2])  # channel 3
    
    # FFT
    fftch1_f = np.abs(fft(ch1_f, N)[:N//2+1])**2
    fftch2_f = np.abs(fft(ch2_f, N)[:N//2+1])**2
    fftch3_f = np.abs(fft(ch3_f, N)[:N//2+1])**2
    
    # automatic DF calculation
    df[ind-1, 0] = np.argmax(fftch1_f) / 2048
    df[ind-1, 1] = np.argmax(fftch2_f) / 2048
    df[ind-1, 2] = np.argmax(fftch3_f) / 2048
    
    # POSTPRANDIAL --------------------------------------------------------
    
    # signal loading
    file_name_p = f'ID{ind}_postprandial.txt'
    dat_p = np.loadtxt(file_name_p)

    # EGG filtering
    ch1_p = filtfilt(b, a, dat_p[:2400, 0])  # channel 1
    ch2_p = filtfilt(b, a, dat_p[:2400, 1])  # channel 2
    ch3_p = filtfilt(b, a, dat_p[:2400, 2])  # channel 3
    
    # FFT
    fftch1_p = np.abs(fft(ch1_p, N)[:N//2+1])**2
    fftch2_p = np.abs(fft(ch2_p, N)[:N//2+1])**2
    fftch3_p = np.abs(fft(ch3_p, N)[:N//2+1])**2
    
    # automatic DF calculation
    df[ind-1, 3] = np.argmax(fftch1_p) / 2048
    df[ind-1, 4] = np.argmax(fftch2_p) / 2048
    df[ind-1, 5] = np.argmax(fftch3_p) / 2048

# conversion from Hz to cpm (cycles per minute)
df[:20] *= 60

# manual corrections for false DF values
df[3, 3] = 3.1934  # ID4 postprandial channel 1
df[3, 4] = 3.1348  # ID4 postprandial channel 2
df[3, 5] = 3.1348  # ID4 postprandial channel 3
df[5, 2] = 2.4900  # ID6 fasting channel 3
df[14, 3] = 2.2560  # ID15 postprandial channel 1
df[16, 0] = 2.9592  # ID17 fasting channel 1
df[16, 1] = 3.0469  # ID17 fasting channel 2
df[16, 2] = 3.0762  # ID17 fasting channel 3

# Generate synthetic data for 20 more participants (15 diabetic, 5 non-diabetic)
np.random.seed(42)  # For reproducibility

# Generate synthetic data for 15 diabetic participants
for ind in range(21, 36):
    df[ind-1, :3] = np.random.uniform(2.0, 3.0, 3)  # Fasting
    df[ind-1, 3:] = np.random.uniform(3.0, 4.0, 3)  # Postprandial

# Generate synthetic data for 5 non-diabetic participants
for ind in range(36, 41):
    df[ind-1, :3] = np.random.uniform(2.5, 3.5, 3)  # Fasting
    df[ind-1, 3:] = np.random.uniform(2.5, 3.5, 3)  # Postprandial

# Add labels for diabetic (1) and non-diabetic (0)
labels = np.zeros(40)
labels[20:35] = 1  # Mark the 15 diabetic participants

# Combine df and labels into a single DataFrame
df_df = pd.DataFrame(np.hstack((df, labels.reshape(-1, 1))), columns=[
    'Fasting_Ch1', 'Fasting_Ch2', 'Fasting_Ch3', 'Postprandial_Ch1', 'Postprandial_Ch2', 'Postprandial_Ch3', 'Diabetic'])

# Shuffle and split the dataset into training, validation, and testing sets
train_df, temp_df = train_test_split(df_df, test_size=0.3, random_state=42, stratify=df_df['Diabetic'])
val_df, test_df = train_test_split(temp_df, test_size=0.5, random_state=42, stratify=temp_df['Diabetic'])

# Save the splits to CSV files
train_df.to_csv('train_df.csv', index=False)
val_df.to_csv('val_df.csv', index=False)
test_df.to_csv('test_df.csv', index=False)

# paired-sample t-tests

# for all subjects ID1-ID20
p_values = []
h_values = []
for i in range(3):
    h, p = ttest_rel(df[:20, i], df[:20, i+3])
    p_values.append(p)
    h_values.append(h)

# for subjects with lower body mass index (BMI)
df_low = np.vstack([df[:3, :], df[8, :], df[11, :], df[13, :], df[15:18, :]])
for i in range(3):
    h, p = ttest_rel(df_low[:, i], df_low[:, i+3])
    p_values.append(p)
    h_values.append(h)

# for subjects ID11-ID20 (higher BMI)
df_high = np.vstack([df[3:8, :], df[9:10, :], df[12, :], df[14, :], df[19, :]])
for i in range(3):
    h, p = ttest_rel(df_high[:, i], df_high[:, i+3])
    p_values.append(p)
    h_values.append(h)

# show p-values and test decisions
print("p-values:", p_values)
print("h-values:", h_values)

